Update and rename FS-1068-open-static-classes.md to FS-1068-open-type… (#481)
fsharp/fslang-design: preview/FS-1068-open-type-declaration.md

# F# RFC FS-1068 - Open Type Declaration

The design suggestion [Open static classes](https://github.com/fsharp/fslang-suggestions/issues/383) has been marked "approved in principle".
This RFC covers the detailed proposal for this suggestion.

[Discussion thread](https://github.com/fsharp/fslang-design/issues/352)

* [x] Approved in principle
* [x] Details: [Resolved to Preview](https://github.com/fsharp/fslang-design/issues/352)
* [x] Implementation: [Complete to Preview](https://github.com/dotnet/fsharp/pull/6325)

## Summary
[summary]: #summary

> Add support for opening types to gain access to static members and nested types, e.g. `open type System.Math`. For example

Reviewer: This isn't defining what opening a type is. It explains what it can do, but not what it is. We're adding a new kind of open declaration to the language - `open type SomeType`, which allows for unqualified access to its static members and nested types. Once a definition like this is in place then this summary is good.

```fsharp
> open type System.Math;;
> Min(1.0, 2.0);;
val it : float = 1.0

> Min(2.0, 1.0);;
val it : float = 1.0
```

## Motivation
[motivation]: #motivation

This greatly increases the expressivity of F# DSLs by allowing method-API facilities such as named arguments, optional arguments and type-directed overloading to be used in the DSL design.

> Type providers can provide static classes, hence this would allow type providers to provide "unqualified" names. This useful for some projections where the natural thing is unqualified, e.g. the "R" type provider always required `R.plot` etc. With this feature, `R.plot` can now be `plot`.

Reviewer: ```suggestion
Type providers can provide static classes, hence this would allow type providers to provide "unqualified" names. This is useful for some projections where the natural thing is unqualified, e.g. the "R" type provider always required `R.plot` etc. With this feature, `R.plot` can now be `plot`.
```

Additionally, important C# DSL APIs are starting to appear that effectively require this.

## Detailed design
[design]: #detailed-design

> Opening of types allows for treating a type somewhat as if it were a module with module-bound functions inside of it. For example:

Reviewer: This is a little too vague. We need to be specific about what is exposed - static members and nested types - and then follow up by saying it is similar to module-bound functions.

```fsharp
type A() =
    static member M(x: int) = x * 2

open type A

M(2)
```

> However, **members are not functions**, so very different rules apply beyond simple cases like the above. This primarily concerns name resolution when overloads are involved, or when considering whether or not we resolve a method or a property first.

Reviewer: > However, **members are not functions**, so very different rules apply beyond simple cases like the above. This primarily concerns name resolution when overloads are involved, or when considering whether or not we resolve a method or a property first.

Neither of these are related to module-bound functions as far as I can tell. You make mention of rules, including the application of some rules in the previous example, but there is no mention of rules for resolving either a module-bound function or a member.

Additionally, what are the concerns? I don't see any description about how we resolve overloads or the resolution of a name of a method or a property. There is an alternatives section describing the construction of a method overload group, but what does it contrast with?

Additionally, visible static fields should also be accessible. Example:

```fsharp
open type System.Data.Common.DbMetaDataColumnNames

printfn "%s" ColumnSize
```

Visible consts should also be allowed. Example:

```fsharp
open type System.Math

PI
```

### Attributes

There are two relevant attributes that are also respected: `AutoOpen` and `RequireQualifiedAccess`.

#### `RequireQualifiedAccess`

As with modules, application of this attribute to a type in F# will require full qualification to use a member defined within.

```fsharp
[<RequireQualifiedAccess>]
type A() =
    static member M(x: int) = x * 2

open type C // Compile error

M(2) // 'M' is not recognized
```

#### `AutoOpen`

> Also like modules, specifying `AutoOpen` on a type automatically brings its members into scope.

Reviewer: Needs mention that the exposed _things_ (not just members) are subject to the rules of name resolution that need defining above.

```fsharp
[<AutoOpen>]
type A() =
    static member M(x: int) = x * 2

M(2) // Can call 'M' without opening 'A'
```

### Opening types reveals nested types

> When opening a .NET type (or a provided type), any accessible nested types also become accessible without qualification.

Reviewer: How do you open a provided type? Can there be an example of getting access to the nested type?

### Extending types

It is possible to use optional type extensions to extend types. Members defined as type extensions are visible when opening these types:

```fsharp
type System.Math with
    static member MyPI = 3.14

open type System.Math

MyPI // works!
```

Resolving members on optional type extensions works the same as with non-static extensions. These rules are not adjusted. That is, members defined in type extensions don't get magically attached to an already-opened static class. The following code will not compile because the `N` member is not in scope:

```fsharp
type A() =
    static member M(x: int) = x * 2

open type A

type A with
    static member N(x: float) = ()

M(2)
N(2.0) // Compile error, not in scope
```

The `open type A` declaration must be placed below the type extension:

```fsharp
type A() =
    static member M(x: int) = x * 2

type A with
    static member N(x: float) = ()

open type C

M(2)
N(2.0) // Works!
```

### Opening types with generic parameter instantiations

> You can open types with generic parameter instantiations:

Reviewer: Can you open it multiple times with different instantiations? e.g.

```fsharp
open type Vector<int>
open type Vector<float>
```

Author: You can, but the latest one would shadow all nested types/method groups from the first one.

```fsharp
open System.Numerics
open type Vector<int>

let x = One // "x" is of type "Vector<int>"
```

But, you cannot have any anonymous parameters:

```fsharp
open System.Numerics
open type Vector<_> // Compile error
```

### Named Types

Named, or nominal, types are only allowed. It means you cannot open a function, tuple, or anonymous record:
```fsharp
open type (int * int) // Compile error, named types are only allowed
```

> Using an abbreviation will still not allow it:

Reviewer: ```suggestion
Using an abbreviation will also not allow it:
```

```fsharp
type MyAbbrev = (int * int)
open type MyAbbrev // Compile error, named types are only allowed
```

> These rules are very similar when using `inherit` in a type definition.

Reviewer: ```suggestion
These rules are very similar to when you use `inherit` in a type definition.
```

open type B

// Both methods are resolved
M(1)
M(1.0)
```

If the signatures for `M` were not unique, then it would not be possible to call `M`. Full qualification is required, or if the source is available, a change to one of the methods is required:

```fsharp
type A() =
    static member M(x: int) = x * 2

type B() =
    static member M(x: int) = x * 2

open type A
open type B

// Error: ambiguous call
M(1)
```

Because overloads coming from multiple opened types are not resolved in the context of a type, we would not require compatibility with the existing rules for resolution.

Because APIs utilizing static members generally use methods instead of properties, we prefer methods over properties and adjust the previous ordered-list as such:

Try to resolve `member-ident` to one of the following, in order:

1. A union case.
2. **A method group.**
3. A property group.
4. A field.
5. An event.
6. **A method group of extension members, by consulting the `ExtensionsInScope` table.**
7. A property group of extension members , by consulting the `ExtensionsInScope` table.
8. A nested type `type-nested`. Recursively resolve .rest if it is present, otherwise return `type-nested`

> That is to say, we will resolve methods over properties in the context of opening static classes or static members extending a static class.

Reviewer: This section reads as if we're going to do this. Should either remove it or reword it to say that it's the avenue we _would_ take.

Additionally, important C# DSL APIs are starting to appear that effectively require this.

## Detailed design
[design]: #detailed-design

Opening of types allows for treating a type somewhat as if it were a module with module-bound functions inside of it. For example:

```fsharp
type A() =
    static member M(x: int) = x * 2

open type A

M(2)
```

> 

Reviewer: Needs explanation of the syntax `open type` and how this is a new kind of open declaration.

Author: Was the summary not sufficient?

Reviewer: There's no formal description of the new syntax form. We'll need that, similar to what was done with `nameof`.

Reviewer 2: I adjusted the test.  It's not very formal still but I think it's what we need for now

We have resolved to use `open type C`.

Issues to do with potential breaking changes do to changes in shadowing led to a change of heart, see https://github.com/fsharp/fslang-design/issues/352#issuecomment-522533863

See original discussion here: https://github.com/fsharp/fslang-design/issues/352#issuecomment-499146012.  

### Open non-static-classes

In 'preview', we could only open static classes and not any class or type. 

Issues to do with potential breaking changes coming from changes in shadowing led to a change of heart, see https://github.com/fsharp/fslang-design/issues/352#issuecomment-522533863

> See original discussion here: https://github.com/fsharp/fslang-design/issues/352#issuecomment-499146012.  

Reviewer: This can just be deleted now, since the updated design will ship as preview, thus obviating this whole section

#### Combining overloaded methods from different types

> We will **not** allow combination of method overloads from different types according to C# rules. The reasoning is for interactivity; we need to be able to shadow method groups in order to not constantly have old method overloads exist.

Reviewer: ```suggestion
We will **not** allow combination of method overloads from different types according to C# rules. The reasoning is for interactivity scenarios such as FSI or Jupyter Notebooks: we need to be able to shadow method groups in order to not constantly have old method overloads exist.
```

That is to say, we will resolve methods over properties in the context of opening static classes or static members extending a static class.

> ## Compatibility

Reviewer: Should have a section on tooling compatibility here. What are implications in tooling for this change, and what needs/needed to be done to ensure a good experience?
